src/combinatorics/quiver/max_flow_min_cut.lean

```diff
@@ -0,0 +1,812 @@
+/-
+Copyright (c) 2023 Aleksandar Milchev, Leo Okawa Ericson, Viggo Laakshoharju. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aleksandar Milchev, Leo Okawa Ericson, Viggo Laakshoharju
+-/
+
+
+import algebra.big_operators.order
+import data.real.basic
+import data.set.finite
+import tactic.induction
+
+/-!
+# Max flow Min cut theorem
+
+In this file we will prove the max-flow min-cut theorem,
+stating that if a maximum flow exists in a flow network,
+then its value is equal to the capacity of a minimum cut in the same network.
+
+## Main results
+
+- `weak_duality`        :
+  The value of every flow is less than or equal to the capacity of every cut.
+
+  direct consequences   :
+  - The value of a max flow is always less than or equal to the capacity of a min cut.
+  - `max_flow_criterion`:
+    If a flow value is equal to a cut capacity in the same flow network, then the flow is maximum.
+  - `min_cut_criterion` :
+    If a capacity of a cut is equal to a flow value in the same network, then the cut is minimum.
+
+- `no_augm_path`        :
+  If the active flow gives a maximum flow, then there is no augmenting path in the residual network.
+- `existence_of_a_cut`  :
+  there exists a cut with a capacity equal to the value of the maximum flow.
+- `max_flow_min_cut`    :
+  If f is a max flow and c is a min cut in the same network, then their values are equal.
+
+The `max_flow_min_cut` lemma is the statement of the max-flow min-cut theorem.
+
+## Notation
+
+- Type V is used for the set of all vertices in our graph. It is a finite type.
+
+## References
+
+- Some of the structure ideas and lemmas can be seen in https://github.com/Zetagon/maxflow-mincut.
+
+-/
+
+open finset
+
+open_locale big_operators
+open_locale classical
+
+notation ` V' ` := univ -- The universe, containing all vertices.
+
+/-!
+  We now define our flow network. Our goal will be to prove weak_duality.
+  As direct consequences, `max_flow_criterion` and `min_cut_criterion` will be proven.
+-/
+
+structure digraph (V : Type*) [inst : fintype V] :=
+  (is_edge : V → V → Prop)
+  (nonsymmetric : ∀ u v : V, ¬ ((is_edge u v) ∧ (is_edge v u)))
+
+structure capacity (V : Type*) [inst : fintype V]
```

Comment:
To me this looks like it would work better as
```suggestion
structure capacity {V : Type*} [inst : fintype V] (g : digraph V)
```
Which says "a capacity on a particular graph". I assume you don't ever need to compare capacities between different graphs?